lib/importers/daily_calls/tp/retriever.rb

```diff
@@ -0,0 +1,31 @@
+require 'mail_retriever'
+
+module Importers
+  module DailyCalls
+    module TP
+      class Retriever
+        class MissingProcessingBlock < StandardError; end
+
+        SEARCH_KEYS = ['SUBJECT', 'TP Daily Call Data'].freeze
+        FILE_NAME_REGEXP = /Daily Data File.*\.xlsx/
+
+        def initialize(config:, mail_retriever: MailRetriever)
+          @config = config
+          @mail_retriever = mail_retriever
+        end
+
+        def process_emails
+          raise MissingProcessingBlock unless block_given?
+
+          retriever = @mail_retriever.new(config: @config)
+          emails = retriever.search(search_keys: SEARCH_KEYS, file_name_regexp: FILE_NAME_REGEXP)
+
+          emails.each do |email|
+            yield email
+            retriever.archive(uid: email.uid)
```

Comment:
The import itself is idempotent

I have wrapped this in a rescue block with bugsnag. I have also modified the saving code so that it wraps the save an AR transaction. That way an error in the file will result in the file being left to process at a future date - without storing partial/invalid data.